lms/djangoapps/bulk_email/messages.py

```diff
@@ -0,0 +1,84 @@
+"""
+Module to define email message related classes and methods
+"""
+from abc import ABC, abstractmethod
+
+from django.contrib.auth import get_user_model
+from django.core.mail import EmailMultiAlternatives
+from edx_ace import ace
+from edx_ace.recipient import Recipient
+
+from lms.djangoapps.bulk_email.message_types import BulkEmail
+from openedx.core.lib.celery.task_utils import emulate_http_request
+
+User = get_user_model()
+
+
+class CourseEmailMessage(ABC):
+    """
+    Abstract base class for course email messages
+    """
+
+    @abstractmethod
+    def send(self):
+        """
+        Triggers sending of email message
+        """
+
+
+class DjangoEmail(CourseEmailMessage):
+    """
+    Email message class to send email directly using django mail API.
+    """
+    def __init__(self, connection, course_email, email_context):
+        """
+        Construct message content using course_email model and context
+        """
+        self.connection = connection
+        template_context = email_context.copy()
+        # use the CourseEmailTemplate that was associated with the CourseEmail
+        course_email_template = course_email.get_template()
+
+        plaintext_msg = course_email_template.render_plaintext(course_email.text_message, template_context)
+        html_msg = course_email_template.render_htmltext(course_email.html_message, template_context)
+
+        # Create email:
+        message = EmailMultiAlternatives(
+            course_email.subject,
+            plaintext_msg,
+            email_context['from_address'],
+            [email_context['email']]
+        )
+        message.attach_alternative(html_msg, 'text/html')
+        self.message = message
+
+    def send(self):
+        """
+        send email using already opened connection
+        """
+        self.connection.send_messages([self.message])
+
+
+class ACEEmail(CourseEmailMessage):
```

Comment:
I'm no expert on the use of `edx-ace`, but I am used to seeing ACE messages inheriting from a common base message class named `BaseMessageType`. Example: https://github.com/openedx/edx-platform/blob/6042972235d088f9afdd62f4b5804d0df2d3a3cc/lms/djangoapps/verify_student/message_types.py#L7.

It's not apparent to me why we used a different (abstract) base class. Why is that?

Reply:
Is it to have a common `send()` function between the DjangoEmail and ACE implementations?

Reply:
Yes to achieve polymorphism or have single method to send both types of emails.

Reply:
`BaseMessageType` is limited to defining message types of edx-ace emails whereas `CourseEmailMessage` is base class for two types of course emails we have 
1. Emails using Custom templates
2. Emails using edx-ace templates